### ISSUE TITLE
[BUG] EAGLE + DP Attention + PD Disaggregation: Deadlock when `index_share_for_mtp_iteration` is enabled for GLM-5.2

### ISSUE THREAD

**Issue by Xavier1994** [sgl-project/sglang#32527]
# EAGLE + DP Attention + PD Disaggregation: Deadlock when `index_share_for_mtp_iteration` is enabled

## Summary

When running EAGLE speculative decoding with DP attention in PD disaggregation decode mode, the server deadlocks during warmup if the model config has `index_share_for_mtp_iteration=True` (e.g., `zai-org/GLM-5.2-FP8`). The deadlock occurs at the GPU level due to asymmetric `can_cuda_graph` decisions between idle and non-idle DP ranks in the `draft()` method, causing NCCL `all_gather` buffer mismatches across ranks.

Removing either `--speculative-algorithm EAGLE` or `--enable-dp-attention` or setting `index_share_for_mtp_iteration=False` in the model config avoids the deadlock.

## Environment

- **Model**: `zai-org/GLM-5.2-FP8` (DeepseekV2 architecture, MoE, FP8 quantization, `index_share_for_mtp_iteration=True`, `index_topk` set)
- **GPU**: 8x B30Z (Blackwell, SM100+)
- **Config**: TP=8, DP=8, `enable_dp_attention=True`, `disaggregation_mode=decode`, `speculative_algorithm=EAGLE`, `speculative_num_steps=3`, `speculative_eagle_topk=1`, `speculative_num_draft_tokens=4`
- **Prefill worker**: Does NOT enable EAGLE (`--speculative-algorithm` not set)

### Decode worker launch command

```bash
python -m sglang.launch_server \
    --model-path zai-org/GLM-5.2-FP8 \
    --disaggregation-mode decode \
    --tp-size 8 --dp-size 8 \
    --enable-dp-attention \
    --speculative-algorithm EAGLE \
    --speculative-num-steps 3 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 4 \
    --cuda-graph-max-bs 128 \
    --watchdog-timeout 100 \
    ... # other args
```

### Symptom

The server starts, captures CUDA graphs, then enters PD disaggregation warmup. During warmup, a request is sent to one DP rank while the other 7 DP ranks are idle. The server hangs and eventually triggers a watchdog timeout.

### Stack trace (py-spy dump)

**The non-idle DP rank (has the request)** — stuck in `draft_forward` via **eager path**:

```
draft_forward (eagle_worker_v2.py:545)
draft (eagle_worker_v2.py:544)
forward_batch_generation (eagle_worker_v2.py:1197)
run_batch (scheduler.py:3351)
```

The eager forward reaches MoE router computation (`dsv3_router_gemm`) and blocks on the NCCL collective.

**The 7 idle DP ranks** — stuck in `verify` via **cuda graph path**:

```
seqlens_expand_triton (pad.py:381)
_apply_cuda_graph_metadata (dsa_backend.py:1533)
init_forward_metadata_out_graph (dsa_backend.py:750)
load_batch (decode_cuda_graph_runner.py:1188)
eagle_prepare_for_verify (eagle_utils.py:558)
run_eagle_verify (eagle_worker_common.py:451)
verify (eagle_worker_v2.py:1504)
forward_batch_generation (eagle_worker_v2.py:1200)
```

These ranks have already completed `draft()` via cuda graph replay and are waiting in `verify()` for the non-idle rank to catch up.

### Background: `seed_dsa_topk_from_draft_extend`

When the model config has `index_share_for_mtp_iteration=True`, the EAGLE worker sets `seed_dsa_topk_from_draft_extend=True`. This flag enables seeding the DSA (Dynamic Sparse Attention) top-k indices from the draft-extend output, so that subsequent draft-decode iterations can reuse the computed indices instead of recomputing.

### The problematic code

In `eagle_worker_v2.py`, the `draft()` method contains:

```python
if (
    can_cuda_graph
    and self.seed_dsa_topk_from_draft_extend
    and not forward_batch.forward_mode.is_idle()
    and draft_input.dsa_topk_indices is None
):
    can_cuda_graph = False  # ← forces eager path
```

This condition downgrades `can_cuda_graph` to `False` when `dsa_topk_indices is None`, forcing the rank onto the **eager path** instead of the **cuda graph path**.

### How the divergence happens

In PD disaggregation decode mode:

1. **Prefill worker does NOT enable EAGLE** — it doesn't run `_draft_extend_for_prefill`, so `req.output_dsa_topk_indices` is never set (remains `None`).

2. **Decode worker** receives the request via PD transfer. In `build_eagle_disagg_draft_input()` (`eagle_disaggregation.py`), `dsa_topk_indices` is assembled from `req.output_dsa_topk_indices`. Since all are `None`, the result is `None`.

3. When a request arrives on **one DP rank** (non-idle) while **other DP ranks are idle**:
   - **Non-idle rank**: `forward_mode.is_idle()` → `False`, so `not is_idle()` → `True`. Combined with `dsa_topk_indices is None`, the condition triggers → `can_cuda_graph = False` → **eager path**.
   - **Idle ranks**: `forward_mode.is_idle()` → `True`, so `not is_idle()` → `False`. The entire condition is skipped → `can_cuda_graph` stays `True` → **cuda graph path**.

4. **Path divergence**: non-idle rank runs `draft_forward` via eager (no DP padding), idle ranks run draft via cuda graph replay (with DP padding to `max(global_num_tokens)`).

5. **Deadlock**: The MoE MLP layer requires all DP ranks to synchronize via NCCL `all_gather`. The eager path uses per-rank token counts (no padding), while the cuda graph path pads all ranks to the same `max_len`. The mismatched buffer sizes cause a GPU-level deadlock on the NCCL collective.

### Why disabling `index_share_for_mtp_iteration` fixes it

When `index_share_for_mtp_iteration=False`:
- `seed_dsa_topk_from_draft_extend = False`
- The condition never triggers
- All ranks consistently use `can_cuda_graph` as determined by `can_run_graph()` (which accounts for DP via `can_run_dp_cuda_graph`)
- No path divergence, no deadlock

## Solution

Instead of downgrading to eager path when `dsa_topk_indices is None`, set a dummy zero-filled tensor. This keeps `can_cuda_graph = True` for all ranks, ensuring they all use the cuda graph path consistently.

### Patch

```diff
--- a/sglang/srt/speculative/eagle_worker_v2.py
+++ b/sglang/srt/speculative/eagle_worker_v2.py
@@ -513,7 +513,11 @@
             and not forward_batch.forward_mode.is_idle()
             and draft_input.dsa_topk_indices is None
         ):
-            can_cuda_graph = False
+            draft_input.dsa_topk_indices = torch.zeros(
+                (draft_input.hidden_states.shape[0], self.dsa_index_topk),
+                dtype=torch.int32,
+                device=draft_input.hidden_states.device,
+            )
```

### Why this works

- The dummy zero tensor makes `dsa_topk_indices is not None`, so the condition no longer triggers, and `can_cuda_graph` stays `True` for the non-idle rank.
- Inside the cuda graph runner's `execute()` method, when `dsa_topk_indices` is not `None`, it copies the seed into the static buffer (`buffers.dsa_seed_topk[:raw_bs].copy_(seed)`). A zero-filled seed is equivalent to no seed — the `draft_forward()` method detects that the seed is all-zeros (via the `seed_dsa_topk_from_draft_extend and dsa_topk_indices is not None` check at line ~618) and sets `spec_info.dsa_topk_indices = None` to trigger recomputation.
- All DP ranks (idle and non-idle) now consistently use the cuda graph path, which includes proper DP padding via `DpPaddingMode.MAX_LEN`, ensuring NCCL `all_gather` buffer sizes match across ranks.

### Verification

The fix was tested with the original launch command. PD disaggregation warmup completes successfully:

```
[2026-07-27 20:30:40] Disaggregation warmup requests completed for all 8 DP ranks
[2026-07-27 20:30:40] End of disaggregation warmup
[2026-07-27 20:30:40] The server is fired up and ready to roll!

**Comment by kpham-sgl** [comment]
> Prefill worker does NOT enable EAGLE — it doesn't run _draft_extend_for_prefill, so req.output_dsa_topk_indices is never set (remains None).

Hi @Xavier1994, in PD Disagg + Spec, we should also enable Spec in the Prefill worker

**Comment by weibingo** [comment]
@Xavier1994 just PD Disaggregation is block ?  can you summit the PR ?
